onnx/defs/tensor/defs.cc

```diff
@@ -2538,164 +2538,178 @@ ONNX_OPERATOR_SET_SCHEMA(
-                  N_H_W_2_seq = SequenceConstruct (N, H, W, int_two_1d)
-                  N_H_W_2 = ConcatFromSequence <axis: int=-1, new_axis: int=0> (N_H_W_2_seq)
-                  grid_2d_then_ = Reshape (grid_N_HW_2, N_H_W_2)
-                  grid_2d_then = CastLike (grid_2d_then_, theta)
-                  },
-              else_branch = g2 () => (grid_3d_else) { # => (float[N, D, H, W, 3])
-                  int_one = Constant <value_int: int=1> ()
-                  minus_one = Constant <value = float {-1.0}> ()
-                  zero = Constant <value = float {0.0}> ()
-                  one = Constant <value = float {1.0}> ()
-                  two = Constant <value = float {2.0}> ()
+                  size_NCDHW_then = Concat <axis=0> (N, C, one_v, H, W)
+              },
+              else_branch = g2 () => (size_NCDHW_else) { # => (int64[5] = [N, C, D, H, W])
+                  # size_NCDHW_else = Identity (size)
+                  int_five = Constant <value_int: int=5> ()
                   N, C, D, H, W = Split <num_outputs: int=5> (size)
-                  int_two_1d = Constant <value_ints=[2]> ()
-                  int_three_1d = Constant <value_ints=[3]> ()
-                  int_five_1d = Constant <value_ints=[5]> ()
-                  constant_D_H_W_shape = Slice (size, int_two_1d, int_five_1d) # [N, C, D, H, W] => [D, H, W]
-                  zeros_D_H_W = ConstantOfShape (constant_D_H_W_shape)
-                  ones_D_H_W = Add (zeros_D_H_W, one)
-
-                  D_float = CastLike (D, zero)
-                  H_float = CastLike (H, zero)
-                  W_float = CastLike (W, zero)
-                  start_d, step_d, start_h, step_h, start_w, step_w = If (constant_align_corners_equal_zero) <
-                      then_branch = h1 () => (start_d_then, step_d_then, start_h_then, step_h_then, start_w_then, step_w_then) { # => (float, float, float, float, float, float)
-                          step_d_then = Div (two, D_float)
-                          step_h_then = Div (two, H_float)
-                          step_w_then = Div (two, W_float)
-
-                          step_d_half = Div (step_d_then, two)
-                          start_d_then = Add (minus_one, step_d_half)
-
-                          step_h_half = Div (step_h_then, two)
-                          start_h_then = Add (minus_one, step_h_half)
-
-                          step_w_half = Div (step_w_then, two)
-                          start_w_then = Add (minus_one, step_w_half)
+                  size_NCDHW_else = Concat <axis=0> (N, C, D, H, W)
+              }
+          >
+          N, C, D, H, W = Split <num_outputs: int=5> (size_NCDHW)
+
+          theta_3d = If (condition_is_2d) <
+              then_branch = g3 () => (theta_then) { # theta: N by 2 by 3 => N by 3 by 4
+                  # use of thetaN23 is a way to make shape inference happy when theta is N by 3 by 4.
```

Comment:
I wonder if it would be a bit simpler and more efficient to just split once, and do two concats? We need one concat to add a row, and one concat to add a column (to convert the 2x3 theta matrix into 3x4), and one split because the new column has to be inserted in the middle (after first 2 columns, and before 3rd column). Why split into every individual element?  Further, GatherElements is probably less efficient as well (in comparison to a simple split).

Reply:
GatherElements is the only way I could found to avoid triggering shape inference errors. Use of Concat to construct 3by4 from 2by3 will fail if theta is already 3by4.

Reply:
Ok. The example with concat might be a good test-case for fixing/extending shape-inference.